flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/ExecNodeMetadataUtil.java

```diff
@@ -0,0 +1,262 @@
+        execNodes.add(StreamExecLocalWindowAggregate.class);
+        execNodes.add(StreamExecLookupJoin.class);
+        execNodes.add(StreamExecMatch.class);
+        execNodes.add(StreamExecMiniBatchAssigner.class);
+        execNodes.add(StreamExecOverAggregate.class);
+        execNodes.add(StreamExecPythonCalc.class);
+        execNodes.add(StreamExecPythonCorrelate.class);
+        execNodes.add(StreamExecPythonGroupAggregate.class);
+        execNodes.add(StreamExecPythonGroupWindowAggregate.class);
+        execNodes.add(StreamExecPythonOverAggregate.class);
+        execNodes.add(StreamExecRank.class);
+        execNodes.add(StreamExecSink.class);
+        execNodes.add(StreamExecSortLimit.class);
+        execNodes.add(StreamExecTableSourceScan.class);
+        execNodes.add(StreamExecTemporalJoin.class);
+        execNodes.add(StreamExecTemporalSort.class);
+        execNodes.add(StreamExecUnion.class);
+        execNodes.add(StreamExecValues.class);
+        execNodes.add(StreamExecWatermarkAssigner.class);
+        execNodes.add(StreamExecWindowAggregate.class);
+        execNodes.add(StreamExecWindowDeduplicate.class);
+        execNodes.add(StreamExecWindowJoin.class);
+        execNodes.add(StreamExecWindowRank.class);
+        execNodes.add(StreamExecWindowTableFunction.class);
+    }
+
+    static {
+        for (Class<? extends ExecNode<?>> execNodeClass : execNodes) {
+            addToLookupMap(execNodeClass);
+        }
+    }
+
+    public static Set<Class<? extends ExecNode<?>>> execNodes() {
+        return execNodes;
+    }
+
+    public static Class<? extends ExecNode<?>> retrieveExecNode(String name, int version) {
+        return lookupMap.get(new ExecNodeNameVersion(name, version));
+    }
+
+    @VisibleForTesting
+    static void addTestNode(Class<? extends ExecNode<?>> execNodeClass) {
+        addToLookupMap(execNodeClass);
+    }
+
+    @SuppressWarnings("rawtypes")
+    private static List<ExecNodeMetadata> extractMetadataFromAnnotation(
+            Class<? extends ExecNode> execNodeClass) {
+        List<ExecNodeMetadata> metadata = new ArrayList<>();
+        ExecNodeMetadata annotation = execNodeClass.getDeclaredAnnotation(ExecNodeMetadata.class);
+        if (annotation != null) {
+            metadata.add(annotation);
+        }
+
+        ExecNodeMetadatas annotations =
+                execNodeClass.getDeclaredAnnotation(ExecNodeMetadatas.class);
+        if (metadata.isEmpty()) {
+            if (annotations != null) {
+                for (ExecNodeMetadata annot : annotations.value()) {
+                    if (annot != null) {
+                        metadata.add(annot);
+                    }
+                }
+            }
+        } else {
+            if (annotations != null) {
+                throw new IllegalStateException(
+                        String.format(
+                                "ExecNode: %s is annotated both with %s and %s. This is a bug, please contact developers.",
```

Comment:
for the future, let's use `this is a bug, please file an issue.`. I don't wanna have mails in my private inbox :D